# LPA - Prática - Aula 23 - 4.3 Resolução do exercício

---
## Parte 1 - Importação das bibliotecas

In [3]:
# Biblioteca Pandas - Manipulação de Dados
import pandas as pd
# Biblioteca Seaborn - Criação de Gráficos
import seaborn as sns
# Biblioteca Matplotlib - Criação de Gráficos
import matplotlib.pyplot as plt

# Biblioteca para ignorar mensagens de warning ao rodas uma célula de código
import warnings
warnings.filterwarnings('ignore')

---
## Parte 2 - Importação dos dados

In [37]:
# Função para importar dos dados da pasta do computador
dados = pd.read_csv('dados_salarial_funcionarios.csv')

In [38]:
# Listando o nome das colunas
dados.columns

Index(['Idade', 'Genero', 'Nivel Ensino', 'Area Formacao', 'Setor', 'Nivel',
       'Faixa Salarial', 'Anos experiencia', 'Faixa Salarial – numero',
       'Data ultimo aumento'],
      dtype='str')

In [39]:
# Imprimindo as 5 primeiras linhas para verificação dos dados importados
dados.head()

,Idade,Genero,Nivel Ensino,Area Formacao,Setor,Nivel,Faixa Salarial,Anos experiencia,Faixa Salarial – numero,Data ultimo aumento
0,38,Masculino,Pós-graduação,Química / Física,Marketing,Gestor,de R$ 4.001/mês a R$ 6.000/mês,Mais de 10 anos,6000,02/01/21
1,39,Masculino,Pós-graduação,Economia/ Administração / Contabilidade / Fina...,Consultoria,Gestor,de R$ 6.001/mês a R$ 8.000/mês,de 2 a 3 anos,8000,06/04/21
2,30,Masculino,Pós-graduação,Computação / Engenharia de Software / Sistemas...,Indústria,Gestor,de R$ 8.001/mês a R$ 12.000/mês,de 1 a 2 anos,12000,15/04/23
3,38,Feminino,Pós-graduação,Outras,Serviços,Gestor,de R$ 12.001/mês a R$ 16.000/mês,Mais de 10 anos,16000,04/06/21
4,36,Masculino,Pós-graduação,Outras Engenharias,Varejo,Gestor,de R$ 8.001/mês a R$ 12.000/mês,de 1 a 2 anos,12000,22/03/23


In [40]:
# Observando o número de linhas e colunas
dados.shape

(1385, 10)

##### Resumo dos dados: 1385 linhas e 10 colunas

---
## Parte 3: Análise Exploratória dos Dados (AED)


### Ter uma primeira visão sobre os dados e extrair os primeiros insights

In [41]:
# Imprime o tipo de dado em cada coluna
# str = variáveis categóricas (texto)
# float e int = variáveis numéricas
dados.dtypes

Idade                      int64
Genero                       str
Nivel Ensino                 str
Area Formacao                str
Setor                        str
Nivel                        str
Faixa Salarial               str
Anos experiencia             str
Faixa Salarial – numero    int64
Data ultimo aumento          str
dtype: object

In [42]:
# Aqui iniciei a depuração nos dados. A coluna "Data ultimo aumento" está como uma str, porém ela deve estar como data
# Utilizarei a função to_datetime do pandas para fazer essa alteração
dados['Data ultimo aumento']=pd.to_datetime(dados['Data ultimo aumento'])

In [43]:
# Verificando a alteração
dados.head()

,Idade,Genero,Nivel Ensino,Area Formacao,Setor,Nivel,Faixa Salarial,Anos experiencia,Faixa Salarial – numero,Data ultimo aumento
0,38,Masculino,Pós-graduação,Química / Física,Marketing,Gestor,de R$ 4.001/mês a R$ 6.000/mês,Mais de 10 anos,6000,2021-02-01
1,39,Masculino,Pós-graduação,Economia/ Administração / Contabilidade / Fina...,Consultoria,Gestor,de R$ 6.001/mês a R$ 8.000/mês,de 2 a 3 anos,8000,2021-06-04
2,30,Masculino,Pós-graduação,Computação / Engenharia de Software / Sistemas...,Indústria,Gestor,de R$ 8.001/mês a R$ 12.000/mês,de 1 a 2 anos,12000,2023-04-15
3,38,Feminino,Pós-graduação,Outras,Serviços,Gestor,de R$ 12.001/mês a R$ 16.000/mês,Mais de 10 anos,16000,2021-04-06
4,36,Masculino,Pós-graduação,Outras Engenharias,Varejo,Gestor,de R$ 8.001/mês a R$ 12.000/mês,de 1 a 2 anos,12000,2023-03-22


In [44]:
# Irei criar duas variáveis para analisar separadamente os dois tipos de variáveis (categóricas e númericas)
# Facilita a Análise Exploratória dos Dados (AED)
numericas_cols=[col for col in dados.columns if dados[col].dtype != 'str']
categoricas_cols=[col for col in dados.columns if dados[col].dtype == 'str']

In [45]:
# Resumo das variáveis numéricas - Imprime alguns valores de medidas de tendências centrais
dados[numericas_cols].describe()

,Idade,Faixa Salarial – numero,Data ultimo aumento
count,1385.000000,1385.000000,1385
mean,30.779783,9992.779783,2022-06-15 08:25:17.978339
min,18.000000,1000.000000,2021-01-01 00:00:00
25%,26.000000,6000.000000,2021-09-19 00:00:00
50%,29.000000,8000.000000,2022-06-09 00:00:00
75%,34.000000,12000.000000,2023-03-05 00:00:00
max,54.000000,45000.000000,2023-12-28 00:00:00
std,6.869018,7353.675097,NaN


In [46]:
# Fazendo o mesmo processo para as variáveis categóricas
dados[categoricas_cols].describe()

,Genero,Nivel Ensino,Area Formacao,Setor,Nivel,Faixa Salarial,Anos experiencia
count,1385,1385,1361,1385,1385,1385,1385
unique,2,6,9,23,4,13,6
top,Masculino,Graduação/Bacharelado,Computação / Engenharia de Software / Sistemas...,Tecnologia/Fábrica de Software,Pleno,de R$ 8.001/mês a R$ 12.000/mês,de 2 a 3 anos
freq,979,495,588,372,440,295,314


In [47]:
# Verificando se excistem valores faltantes nos dados
# Is NA = Valores faltantes
# Any = Trará um valor booleano
dados.isna().any()

Idade                      False
Genero                     False
Nivel Ensino               False
Area Formacao               True
Setor                      False
Nivel                      False
Faixa Salarial             False
Anos experiencia           False
Faixa Salarial – numero    False
Data ultimo aumento        False
dtype: bool

In [48]:
# Quantificando quantos valores faltantes possuem cada coluna
# Is NA = Valores Faltantes
# Sum = Quantidade
dados.isna().sum()

Idade                       0
Genero                      0
Nivel Ensino                0
Area Formacao              24
Setor                       0
Nivel                       0
Faixa Salarial              0
Anos experiencia            0
Faixa Salarial – numero     0
Data ultimo aumento         0
dtype: int64

In [56]:
# Agora irei substituir os valores faltantes por um texto (Nada consta)
# Na mesma variável, chama dados sem precisar criar uma outra variável
# Fill NA = Preencher os valores faltantes
dados['Area Formacao'] = dados['Area Formacao'].replace(['', None], 'Nada Consta')

In [50]:
# Ou se preferir, posso excluir a coluna inteira se os dados não forem importantes. Usando o seguinte código:
# Axis1 = Sentido da coluna
# Axis2 = Sentido das linhas 
##dados.drop('Area Formacao', axis=1, inplace=True)

In [57]:
# Depois da depuração, verificarei se ainda há valores faltantes
dados.isna().sum()

Idade                      0
Genero                     0
Nivel Ensino               0
Area Formacao              0
Setor                      0
Nivel                      0
Faixa Salarial             0
Anos experiencia           0
Faixa Salarial – numero    0
Data ultimo aumento        0
dtype: int64

In [58]:
# Verificando se há valores duplicados
dados.duplicated().sum()

np.int64(0)

In [60]:
# Para remover os valores duplicados eu posso usar a função
## dados.drop_duplicates(inplace=True)

In [61]:
# Verificando o número de linhas e colunas após as mudanças realizadas. Caso eu utilizasse a função acima, o números
# de linhas seria reduzido
dados.shape

(1385, 10)

In [62]:
# Para uma verificação de distruição de dados, eu utilizo a função .value_counts()
dados['Nivel'].value_counts()

Nivel
Pleno     440
Júnior    409
Sênior    370
Gestor    166
Name: count, dtype: int64

In [63]:
# Posso utilizar a mesma função para visualizar na categoria "Faixa Salarial"
dados['Faixa Salarial'].value_counts()

Faixa Salarial
de R$ 8.001/mês a R$ 12.000/mês     295
de R$ 6.001/mês a R$ 8.000/mês      255
de R$ 4.001/mês a R$ 6.000/mês      242
de R$ 12.001/mês a R$ 16.000/mês    140
de R$ 3.001/mês a R$ 4.000/mês      113
de R$ 2.001/mês a R$ 3000/mês       108
de R$ 1.001/mês a R$ 2.000/mês       79
de R$ 16.001/mês a R$ 20.000/mês     59
de R$ 20.001/mês a R$ 25.000/mês     26
de R$ 25.001/mês a R$ 30.000/mês     22
Menos de R$ 1.000/mês                19
de R$ 30.001/mês a R$ 40.000/mês     18
Acima de R$ 40.001/mês                9
Name: count, dtype: int64

In [ ]:
# Para criar uma nova coluna eu usaria a seguinte função
# Função para calcular a média salarial dos funcionários
# Mean = Função de média
# Round = Função para arredondar
dados['Média Salarial']=dados['Faixa Salarial – numero'].mean().round()

In [67]:
# Resultado da adição da coluna
dados.head()

,Idade,Genero,Nivel Ensino,Area Formacao,Setor,Nivel,Faixa Salarial,Anos experiencia,Faixa Salarial – numero,Data ultimo aumento,Média Salarial
0,38,Masculino,Pós-graduação,Química / Física,Marketing,Gestor,de R$ 4.001/mês a R$ 6.000/mês,Mais de 10 anos,6000,2021-02-01,9993.0
1,39,Masculino,Pós-graduação,Economia/ Administração / Contabilidade / Fina...,Consultoria,Gestor,de R$ 6.001/mês a R$ 8.000/mês,de 2 a 3 anos,8000,2021-06-04,9993.0
2,30,Masculino,Pós-graduação,Computação / Engenharia de Software / Sistemas...,Indústria,Gestor,de R$ 8.001/mês a R$ 12.000/mês,de 1 a 2 anos,12000,2023-04-15,9993.0
3,38,Feminino,Pós-graduação,Outras,Serviços,Gestor,de R$ 12.001/mês a R$ 16.000/mês,Mais de 10 anos,16000,2021-04-06,9993.0
4,36,Masculino,Pós-graduação,Outras Engenharias,Varejo,Gestor,de R$ 8.001/mês a R$ 12.000/mês,de 1 a 2 anos,12000,2023-03-22,9993.0
